src/ast_editor.ml

```diff
@@ -0,0 +1,549 @@
+open Import
+open Ppx_utils
+
+let webview_map = ref (Map.empty (module String))
+
+let original_mode = ref true
+
+let hover_disposable_ref = ref None
+
+let doc_string_uri ~document = Uri.toString (TextDocument.uri document) ()
+
+let pp_doc_to_changed_origin_map = ref (Map.empty (module String))
+
+let origin_to_pp_doc_map = ref (Map.empty (module String))
+
+let entry_exists k d =
+  Map.existsi !origin_to_pp_doc_map ~f:(fun ~key ~data ->
+      String.equal d data && String.equal k key)
+
+let put_keys_into_a_list data_serached map =
+  Map.filteri ~f:(fun ~key:_ ~data -> String.equal data_serached data) map
+  |> Map.to_alist |> List.unzip |> fst
+
+let find_webview_by_doc ~document =
+  match Map.find !webview_map (doc_string_uri ~document) with
+  | wv_opt when !original_mode -> wv_opt
+  | None -> (
+    match
+      put_keys_into_a_list (doc_string_uri ~document) !origin_to_pp_doc_map
+    with
+    | [ k ] -> (
+      match Map.find !webview_map k with
+      | wv_opt when not !original_mode -> wv_opt
+      | _ -> None)
+    | _ -> None)
+  | _ -> None
+
+let remove_keys_by_data map d =
+  map := Map.filteri ~f:(fun ~key:_ ~data -> not (String.equal data d)) !map
+
+let set_changes_tracking origin pp_doc =
+  origin_to_pp_doc_map :=
+    Map.set !origin_to_pp_doc_map
+      ~key:(doc_string_uri ~document:origin)
+      ~data:(doc_string_uri ~document:pp_doc);
+  pp_doc_to_changed_origin_map :=
+    Map.set
+      !pp_doc_to_changed_origin_map
+      ~key:(doc_string_uri ~document:pp_doc)
+      ~data:false
+
+let set_origin_changed ~data ~key =
+  pp_doc_to_changed_origin_map :=
+    Map.set !pp_doc_to_changed_origin_map ~key ~data
+
+let on_origin_update_content changed_document =
+  match
+    Map.find !origin_to_pp_doc_map (doc_string_uri ~document:changed_document)
+  with
+  | Some key -> set_origin_changed ~key ~data:true
+  | None -> ()
+
+let read_html_file () =
+  let filename = Node.__dirname () ^ "/../astexplorer/dist/index.html" in
+  Fs.readFile filename
+
+let document_eq a b =
+  String.equal
+    (Uri.toString (TextDocument.uri a) ())
+    (Uri.toString (TextDocument.uri b) ())
```

Comment:
Can we add `Uri.equal` instead?

Reply:
Added in 22f72393c5d7beb12713a2097a78bd25c5c8a828